File: scripts/extract-db-json.js
const fs = require("fs");

const inputFile = process.argv[2];
const outputFile = process.argv[3];

const content = fs.readFileSync(inputFile, "utf-8");

// Parse the JSON wrapper
const jsonData = JSON.parse(content);
const text = jsonData[0].text;

// Extract data between untrusted markers
const startMarker = /\[{\\"/;
const endMarker = /}\]\n</;

const match = text.match(startMarker);
if (!match) {
  console.error("Could not find data start");
  process.exit(1);
}

const startIdx = match.index;
const dataSection = text.substring(startIdx);

// Find the end
const endMatch = dataSection.match(endMarker);
if (!endMatch) {
  console.error("Could not find data end");
  process.exit(1);
}

const jsonString = dataSection.substring(0, endMatch.index + 2);

// Unescape the JSON
const unescaped = jsonString.replace(/\\"/g, '"').replace(/\\\\/g, "\\");

// Write to output
fs.writeFileSync(outputFile, unescaped);
console.log(`Extracted JSON to ${outputFile}`);

File: scripts/extract-db-json.py
#!/usr/bin/env python3
import json
import sys
import re

with open(sys.argv[1], 'r') as f:
    content = f.read()

# Parse outer JSON
data = json.loads(content)
text = data[0]['text']

# Find the escaped JSON array
match = re.search(r'\[\{\\\"function_name.*?\}\]', text, re.DOTALL)
if not match:
    print("Could not find JSON array", file=sys.stderr)
    sys.exit(1)

escaped_json = match.group(0)

# Unescape the JSON
json_str = escaped_json.replace('\\"', '"').replace('\\\\', '\\')

# Parse and pretty-print
parsed = json.loads(json_str)

with open(sys.argv[2], 'w') as f:
    json.dump(parsed, f, indent=2)

print(f"Extracted {len(parsed)} functions to {sys.argv[2]}")
